api/src/retry.rs

```diff
@@ -0,0 +1,204 @@
+use reqwest::{blocking::Response, Result};
+use std::sync::atomic::{AtomicBool, Ordering::Relaxed};
+use std::thread::sleep;
+use std::time::Duration;
+
+/// Strategy to use if retrying .
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub enum RetryStrategy {
+    /// The first request by the client will not be retried, but subsequent requests will.
+    /// This allows fast failure if the client cannot reach the API endpoint at all, but
+    /// helps to mitigate failure in long-running operations spanning multiple requests.
+    Automatic,
+    /// Always attempt to retry requests.
+    Always,
+}
+
+/// Configuration for the Reinfer client if retrying timeouts.
+#[derive(Clone, Debug, PartialEq)]
+pub struct RetryConfig {
+    /// Strategy for when to retry after a timeout
+    pub strategy: RetryStrategy,
+    /// Maximum number of retries to attempt.
+    pub max_retry_count: u8,
+    /// Amount of time to wait for first retry.
+    pub base_wait: Duration,
+    /// Amount of time to scale retry waits. The wait before retry N is an exponential backoff
+    /// using the formula `wait = retry_wait * (backoff_factor * N)`.
+    pub backoff_factor: f64,
+}
+
+#[derive(Debug)]
+pub(crate) struct Retrier {
+    config: RetryConfig,
+    is_first_request: AtomicBool,
+}
+
+impl Retrier {
+    pub fn new(config: RetryConfig) -> Self {
+        Self {
+            config,
+            is_first_request: AtomicBool::new(true),
+        }
+    }
+
+    pub fn with_retries(&self, f: impl Fn() -> Result<Response>) -> Result<Response> {
+        if self.is_first_request.swap(false, Relaxed)
+            && self.config.strategy == RetryStrategy::Automatic
+        {
+            return f();
+        }
+
+        for i in 0..self.config.max_retry_count {
```

Comment:
nit `i_retry` rather than `i`? this scales much better when nested for loops come into it.

Reply:
Gone for `retry_count`.